File: python/01-arrays-hashing/005-top-k-frequent-elements.py
"""
NeetCode 150 - Problem 5: Top K Frequent Elements

Approach:
1. Use a dictionary to count the frequency of each number.
2. Create a bucket list where index represents frequency.
   Each bucket stores numbers that appear that many times.
3. Traverse the bucket list from highest frequency to lowest
   and collect the top k frequent elements.

Time Complexity: O(n)
Space Complexity: O(n)
"""

from typing import List

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # Step 1: Frequency map
        count = {}
        freq = [[] for _ in range(len(nums) + 1)]

        for n in nums:
            count[n] = 1 + count.get(n, 0)

        # Step 2: Bucket placement
        for n, c in count.items():
            freq[c].append(n)

        # Step 3: Collect top k elements
        res = []
        for i in range(len(freq) - 1, 0, -1):
            for n in freq[i]:
                res.append(n)
                if len(res) == k:
                    return res

        return res

File: java/01-arrays-hashing/005-top-k-frequent-elements.java
/*
 * NeetCode 150 - Problem 5: Top K Frequent Elements
 *
 * Approach:
 * 1. Use a HashMap to count the frequency of each number.
 * 2. Create a bucket array where the index represents the frequency.
 *    Each bucket stores numbers that appear that many times.
 * 3. Traverse the bucket array in reverse (highest frequency first)
 *    and collect the top k frequent elements.
 *
 * This avoids using a heap and achieves linear time complexity.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

import java.util.*;

class Solution {
    public int[] topKFrequent(int[] nums, int k) {

        // Step 1: Count frequencies
        Map<Integer, Integer> frequencyMap = new HashMap<>();
        for (int num : nums) {
            frequencyMap.put(num, frequencyMap.getOrDefault(num, 0) + 1);
        }

        // Step 2: Bucket array (index = frequency)
        @SuppressWarnings("unchecked")
        List<Integer>[] bucket = new List[nums.length + 1];

        for (Map.Entry<Integer, Integer> entry : frequencyMap.entrySet()) {
            int num = entry.getKey();
            int freq = entry.getValue();

            if (bucket[freq] == null) {
                bucket[freq] = new ArrayList<>();
            }

            bucket[freq].add(num);
        }

        // Step 3: Collect top k frequent elements
        int[] result = new int[k];
        int index = 0;

        for (int i = bucket.length - 1; i >= 0 && index < k; i--) {
            if (bucket[i] != null) {
                for (int num : bucket[i]) {
                    result[index++] = num;
                    if (index == k) break;
                }
            }
        }

        return result;
    }
}
